### ISSUE TITLE
imageGW checksum fails when transfers are interrupted

### ISSUE THREAD

**Issue by dmjacobsen** [NERSC/shifter#9]
Originally reported by Miguel Gila (CSCS) on 2015-12-15

If a download is partially done (e.g. when the process gets killed) for a specific image, subsequent pulls of the image fail with "OSError: Pull failed".
A workaround is cleaning up CacheDirectory.
Here is an example:
```
[2015-12-15 14:59:44,917: INFO/MainProcess] Received task: imageworker.dopull[6a3d2d55-20c5-41e8-96d7-af08191592d3]
[2015-12-15 14:59:44,918: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x2b99274977d0> (args:(u'imageworker.dopull', u'6a3d2d55-20c5-41e8-96d7-af08191592d3', [{u'pulltag': u'miguelgila/wlcg_wn:20151214', u'tag': u'miguelgila/wlcg_wn:20151214', u'itype': u'docker', u'system': u'santis'}], {u'TESTMODE': 0}, {u'utc': True, u'is_eager': False, u'chord': None, u'group': None, u'args': [{u'pulltag': u'miguelgila/wlcg_wn:20151214', u'tag': u'miguelgila/wlcg_wn:20151214', u'itype': u'docker', u'system': u'santis'}], u'retries': 0, u'delivery_info': {u'priority': 0, u'redelivered': None, u'routing_key': u'santis', u'exchange': u'santis'}, u'expires': None, u'hostname': 'celery@worker.queue.santis', u'task': u'imageworker.dopull', u'callbacks': None, u'correlation_id': u'6a3d2d55-20c5-41e8-96d7-af08191592d3', u'errbacks': None, u'timelimit': [None, None], u'taskset': None, u'kwargs': {u'TESTMODE': 0}, u'eta': None, u'reply_to': u'83e23994-8fd8-333c-ac07-7d366c5c0796', u'id': u'6a3d2d55-20c5-41e8-96d7-af08191592d3', u'headers': {}}) kwargs:{})
[2015-12-15 14:59:44,919: DEBUG/MainProcess] Task accepted: imageworker.dopull[6a3d2d55-20c5-41e8-96d7-af08191592d3] pid:18565
[2015-12-15 14:59:44,919: DEBUG/Worker-5] dopull system=santis tag=miguelgila/wlcg_wn:20151214
[2015-12-15 14:59:44,922: DEBUG/Worker-5] doing image pull for registry-1.docker.io miguelgila/wlcg_wn 20151214
[2015-12-15 14:59:52,084: WARNING/Worker-5] checksum mismatch, failure
[2015-12-15 14:59:52,088: INFO/Worker-5] Worker: Pull failed
[2015-12-15 14:59:52,089: ERROR/Worker-5] ERROR: dopull failed system=santis tag=miguelgila/wlcg_wn:20151214
[2015-12-15 14:59:52,092: ERROR/MainProcess] Task imageworker.dopull[6a3d2d55-20c5-41e8-96d7-af08191592d3] raised unexpected: OSError('Pull failed',)
Traceback (most recent call last):
  File "/root/shifter/venv_2.7.9/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/root/shifter/venv_2.7.9/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/root/shifter/miguelgila-shifter-54925ed85fe9/imagegw/src/imageworker.py", line 243, in dopull
    raise OSError('Pull failed')
OSError: Pull failed
```

**Comment by dmjacobsen** [comment]
I suppose the right thing to do here is run a checksum if the layer already exists. If it matches, use it, if not delete the existing layer and try downloading again.

**Comment by dmjacobsen** [comment]
Shane responded:
I think we can just put things in a temp file location during the fetch. Then if it fails it should still be incomplete.

**Comment by dmjacobsen** [comment]
this should have been fixed with pr #41
